compiler/rustc_middle/src/ty/consts/kind.rs

```diff
@@ -68,6 +73,104 @@ pub enum ConstKind<'tcx> {
     /// A placeholder for a const which could not be computed; this is
     /// propagated to avoid useless error messages.
     Error(ty::DelaySpanBugEmitted),
+
+    /// Abstract which contains an expression which has partially evaluated items.
+    Abstract(Expr<'tcx>),
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Hash)]
+#[derive(HashStable, TyEncodable, TyDecodable)]
+pub enum Expr<'tcx> {
```

Comment:
please derive `TypeFoldable` and `TypeVisitable` instead

Reply:
This would require implementing it for `&[ty::Const]`, is that fine? That I think would have to be a manual implementation

Reply:
yes, implementing this manually for `&[ty::Const]` is fine. Though actually, we probaby want to use `&'tcx ty::List<ty::Const<'tcx>>` instead of a slice here anyways, because otherwise we end up with a lot of duplicate slices in memory